### ISSUE TITLE
Show which distinct versions of a file exist

### ISSUE THREAD

**Issue by bwurst** [borgbackup/borg#729]
We do backup with attic (as attic is frozen by now, we think about switching to borg) and there's one thing that really bugs me. I cannot see which distinct versions of a specific file exists. I can mount and extract a daily snapshot and the file is there but I cannot see if this file is the same as the day before.

We operate rdiff-backup in parallel mostly because of this feature. It's really useful if a web app has been hacked for example. Normally, files don't change too much, if some files are changed when no upgrade has been done, we pay stronger attention to this. For this to work, we need to know which file has been changed in which snapshot or, other approach, which versions of a file really exists.

Does borg, internally, know this? Is there some kind of file checksum list that can be used for this? I'd do some coding if the data structure does provide this kind of information but I did not find it by now...

***

:moneybag: [there is a bounty for this](https://www.bountysource.com/issues/31735500-show-which-distinct-versions-of-a-file-exist)

**Comment by infectormp** [comment]
+1
To solve this task we currently developing a webapp that will perform function that you described.

**Comment by enkore** [comment]
You can do this by extracting the file(s) of interest and comparing them. E.g. extract the interesting part of snapshot *n* into `a/`, and *n-1* into `b/`. Diff a, b. If not different, move `b/` to `a/`, extract *n-2* to `b/`. Diff again. Repeat as necessary.

**Comment by infectormp** [comment]
@enkore your workaround good only for non regular operations. If you need to view changes constantly it is uncomfortable. You need have a lot of time to make this operations regularly.

**Comment by enkore** [comment]
Yes. Still, this method would be quite usable for investigating data loss, breaches and the like.

**Comment by bwurst** [comment]
@enkore: this does not work for day to day operation. We back up about 300 GB of user's content per machine. Extracting and comparing this won't be fun.
I thought about building a checksum table daily but it's also a huge waste of resources because I think borg does this anyway. Just now I don't have much time so I'll accept that this seems not to be possible out of the box.

**Comment by enkore** [comment]
Borg doesn't know if a file changed relative to the previous archive, because that is not a concept Borg has (it does not know what archive is considered "the previous archive" by the user).

If using the cache then the output of `--list -v` when creating the archive will show which files have been changed *relative to the file cache used*. E.g. `U some_binary` => unchanged. `A sshd` => someone changed sshd.

If no cache is used then I think every file will always show as `A`.

**Comment by bwurst** [comment]
Ah, I see, sorry. I thought, deduplication does mean "only save changed files" as it is in rsync-based backup tools but this is a different approach, I forgot. Does borg save a whole-file checksum? If not, (optinally) generating a file checksum table along with each run could be a great thing. So one can grep for a specific file in this table and then see which copies are the same.

**Comment by enkore** [comment]
Yes, this should be possible, the implementation would be similar to extract, but not the same:

- For every file to compare, fetch the associated item from the archives (A, B) to compare
- Compare whether the chunk lists of the two items match
- If they don't, the file has been changed between A and B

=> This doesn't read the actual data. But if the chunk, uh, "types" (compression) are different, then we'd need to read the chunks and decompress, decrypt them and compare them (e.g. by hashing). This wouldn't necessarily mean writing them to disk, one could just read one chunk after another and update a hash function, so you'd need only a couple MB for any file.

**Comment by manwegit** [comment]
At the moment I use do different solutions in parallel.
1) Store log for each backup: I can easily grep which backups contained change to which file.
2) `borg list /backup/repo --short|xargs -r -d$'\n' -I% bash -c 'borg list  --list-format="{path}{TAB}{mtime:%s}{TAB}{size}{LF}"  /backup/repo::"%" > "/tmp/archive-%"'`
   This is faster than mount/extract + diff/md5sum, but it does not catch all changes.

**Comment by ThomasWaldmann** [comment]
borg does not compute or store full-file hash. it has the hashes/hmacs of all the pieces, though.

**Comment by enkore** [comment]
[enkore:diff-proto](https://github.com/enkore/borg/tree/diff-proto) is a rough sketch of how this could be done.

The major technical problem I see here is that finding a *specific* file is not straightforward in this data structure.

Average run-time might still be made acceptable by being much smarter about how this is done:
Iterate over both archives in parallel, add pattern-matching items to a list-of-tuples `[(refitem, compitem), …] + (None, compitem)` (or `(refitem, None)`, or `(refitem, compitem)`, if both paths are identical in an iteration), compare all completed two-tuples in the list by the proposed 'algorithm'.

**Comment by enkore** [comment]
The naive approach implemented in abbfebab doesn't work for larger archives. I implemented the 'iterating zipper' strategy I described in the comment above, it works fairly well.

These numbers are for comparing two large (~400 GB raw data, >2 million files) archives with small differences (deduplicated, compressed difference about 1.5 GB).

If we need to compare file contents, and I did this by SHA512 in this prototype (the idea is that you could specify other hash algorithms and thus easily validate files in the archive against MD5SUMS/SHA1SUMS/SHA256SUMS/whatever supplied by an upstream), then this takes little more than three minutes wall clock time:

    borg diff backup.borg::2016-03-10-tiger 2016-03-11-tiger  187.79s user 2.62s system 99% cpu 3:11.86 total

If on the other hand we are somehow able to establish that the chunker-params of the two archives in question are identical (here I know this, because I created them) and I don't need the hashes of differing files, then this time shortens drastically:

    borg diff backup.borg::2016-03-10-tiger 2016-03-11-tiger  69.23s user 0.36s system 98% cpu 1:10.30 total

Just a little over one minute.

--

Determining chunker-params: I think these could just be stored in the archive metadata for new versions, this wouldn't break anything (?). Parsing the command line in the metadata could also work... but would not work at all if the defaults in the source were changed.

**Comment by ThomasWaldmann** [comment]
Hmm, feature creep? We started from "show distinct files" and now it also shall list the hashes of these files with configurable hash algorithm to compare against a external database.

**Comment by enkore** [comment]
Maybe, it is a little feature creep. On the other hand the additional code wouldn't be much, nor complex. So the question on whether a "borg diff" subcommand, which may or may not be included with borg in some future version, should have that feature should be more like - do people need it? does it do something that would be much harder otherwise?

I don't need hash-lists. Quickly listing which files changed between archives, that is something I find handy, however. @bwurst above just mentioned using checksum lists for that and paying attention to changed system files when nothing was upgraded, so I included it.

**Comment by ThomasWaldmann** [comment]
Just in case we decide at a time to store the full file hash into file metadata, the feature-creep "use and display configurable hash" would make that stored hash useless for n-1 cases. Otherwise we could just use such a stored hash.

**Comment by enkore** [comment]
You convinced me, it's better not to include it. If hashes are needed, the files of interest can be passed to `extract` and hashed any way the user needs it. Good point.

Another thing we have not discussed yet is how the output should look like. We could just print a list of changed files, one per line (how to handle paths with a `\n` in them?).

Including in what way it changed would make sense (added, removed, different), especially if we'd decide to [optionally] check meta data (owner, group, permissions, [ACLs? xattr?],  time stamps don't really make sense imho, they change all the time).

**Comment by ThomasWaldmann** [comment]
one interesting thing (and there is another ticket about that) is to determine the amount of data change.

there are 2 ways this could be determined:

1. in a diff-like way (that would somehow fit in here): when comparing chunk id list of a file now to the reference file, how many chunk ids were removed, how many added (and then, do not just display that count, but determine the sum of these chunk sizes)?
2. how many unique chunks does the file have (compared to everything else in the repo), this doesn't fit into diff semantics.

**Comment by enkore** [comment]
A very good idea.

    $ time borg diff --chunks backup.borg::2016-03-10-tiger 2016-03-11-tiger 'home/…/INBOX'
    home/…/INBOX different by chunk-compare
            1.00 [=390=±]
    borg diff --chunks backup.borg::2016-03-10-tiger 2016-03-11-tiger
    58.84s user 0.31s system 99% cpu 59.740 total

I do this with difflib's SequenceMatcher. A small re.repl statement replaces overly long repetitions of the same diff symbols with numerals. For small files the output is skipped since it's meaningless.

While SequenceMatcher has poor (theoretical) complexity (cubic), it's still quite fast (we are only working on the chunk ids, and even large files only have a few thousand chunks at most). Enabling this output just costs a couple seconds over an entire run.

    home/…/INBOX different by chunk-compare
            1.00 [=390=±]
    home/…/global-messages-db.sqlite different by chunk-compare
            0.72 [±±=±=30=±=13=±=====±===±=±===±±±====±±±±====±==±=15=±=±±===±±=9=±=13=±±===±=±±±=±===±=±===±±±±=±=±=±=±±=±±±=±=±===±±±==±4±]

So here we see that my INBOX (~1 GB) was only appended to between the two archives. And how this SQLite database was changed alot across the whole file. I don't know what it does, probably a search index built by Thunderbird or something like that.

**Comment by ThomasWaldmann** [comment]
You don't need SequenceMatcher, just a set difference A-B (removed) and B-A (added) and then sum up the sizes.

**Comment by enkore** [comment]
For output like this?

        +18,-17 chunks, 19 total
        Estimated size: 39 MB

I dunno. I think the "diff-bar" gives a better/quicker idea of what has happened in the file.

**Comment by ThomasWaldmann** [comment]
-xx MB, +xx MB (or -<removedfromold>/<oldsize>, +<addedtonew>/<newsize> ?).

And I don't think we should go into trying to analyze what has happened in the file, no diff bar.

Granularity is ~1MB, so a changed chunk could mean a 1 or 1000000 bytes change. If someone wants to know what has changed precisely, he still can run diff (or another tool) on a fuse mount.

**Comment by ThomasWaldmann** [comment]
just to keep an idea about "show versions of a file":

if we would process all archives in a repo when doing a fuse mount (might take a while), we could create a directory for any file that ever existed. inside that directory, files named like a hash could be created for every "version" of the file.

the hash would be H(chunkids). the mtime of this "hash file" would be the mtime from the archive, same for the size. this way, there are no "collisions" (same name, different content) and the hash is efficient to compute.

**Comment by ThomasWaldmann** [comment]
PR #1129 \o/

**Comment by anarcat** [comment]
> if we would process all archives in a repo when doing a fuse mount (might take a while), we could create a directory for any file that ever existed. inside that directory, files named like a hash could be created for every "version" of the file.

that looks like the moinmoin revision storage. :) i guess if this is optional it's okay, but otherwise it would look pretty weird... is there a versionning layer in FUSE that could be accessed?

**Comment by ThomasWaldmann** [comment]
I didn't think about MoinMoin, I promise. ;)

And aside from using a directory to keep together versions of same file, there is not that much similarity. Esp. the disambiguation is "hash" based, not just a counter.

I don't know about a FUSE version layer (or how that should even work in tools above the fuse layer).

**Comment by anarcat** [comment]
I looked a bit more into this: there are several versionned filesystems out there [according to wikipedia](https://en.wikipedia.org/wiki/Versioning_file_system), the most proeminent being Git, which doesn't quite apply to our use case, obviously, unless we want to implement a git compatibility layer on top of borg, which could be interesting, but there's `bup` that already does that, somewhat.

There's also [wayback](http://wayback.sourceforge.net/) which seems to be abandoned and doesn't explicitly state on their homepage how previous versions are accessed.

Then there's the more recent [fuse-copyfs](https://github.com/cognusion/fuse-copyfs) that creates new copies of files on write, and keeps history. The way they implement version discovery is with a special tool:

```
cpy-fs $ copyfs-fversion somedir/file-1
File file-1 ('*' is active) :
  v1.0  : -rw-r--r--  widan    users             3 Fri Dec 10 14:21:41 2004
  v2.0  : -rw-r--r--  widan    users             3 Fri Dec 10 14:21:43 2004 [*]
  v3.0  : -rw-r--r--  widan    users             3 Fri Dec 10 14:21:45 2004
```

This seems like an interesting approach to me: the filesystem could look perfectly normal, but then borg commands could inspect history of a given file (say `borg list-versions` or simply overloading `borg list`). `borg diff` could also be extended to do the right thing with specific version markers (e.g. `file@archive`).

This could be implemented on top of the normal fuse mount, without modifying the file layout. For a more hands-on direct approach, the multiple files per directory approach of #1129 could still be implemented as a toggle...

**Comment by ThomasWaldmann** [comment]
@anarcat: If you want the "normal view", there will be the option to just not use the "versioned view". The is way simpler than what you propose above.

**Comment by anarcat** [comment]
i guess my point is "i want both" and i think that was more or less the request above. the original post was refering to rdiff-backup, which, if i remember correctly, has a file structure like:

```
file
file.snapshots/version0
file.snapshots/version1
```

where `file` is the latest version, and previous versions are stored under a `file.snapshot` subdirectory. that means that the file layout is preserved and current files are in their expected locations.

**Comment by ThomasWaldmann** [comment]
not sure if it is really better if we merge both views into what you propose above (and only offer that).

e.g. if you'ld use "find" or any other recursive command on that merged view, it would go through all versions that ever existed. If you do not want that, this is a serious performance problem. If you want that, you could also use the "versioned view" as it is now.

**Comment by anarcat** [comment]
i am not saying we should offer only that "rdiff-backup-like" view. I am saying the "versionned view" should be like that so that *some* normal semantics remain.

the "non-versionned view" should remain in any case, I agree. furthermore, i think it would be great to be able to inspect versions using commandline tools, without resorting (re)mounting with fuse, something which can be expensive for a one off task.

**Comment by DrupaListo-com** [comment]
alraban just supported this issue with 100$ - https://www.bountysource.com/issues/31735500-show-which-distinct-versions-of-a-file-exist/backers

I am wondering is their bot gonna post here ?

**Comment by ThomasWaldmann** [comment]
@DrupaListo-com Due to security concerns, I did not give bountysource write access to the project.

Thanks to alraban for supporting this!

**Comment by ThomasWaldmann** [comment]
About the state or PR #1129: it is currently blocked because there were such a lot of ideas about how to reflect the file versions in the FUSE file system and it was not clear to me yet which way to go.

**Comment by alraban** [comment]
@ThomasWaldmann, I'm just glad to be able to support the project a bit.
I look forward to seeing whatever implementation you all settle on.  For my part, as long as there's an option for a versioned and a non-versioned view, I'll be happy.
